### PR TITLE
future3 - Feature "sync shared"

### PR DESCRIPTION
**Pull request by AlvinSchiller**
Im adding the Feature "sync shared" (https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/1978) from v2 to future3.

So far i got my code working ([components/synchronisation/sync_shared](https://github.com/AlvinSchiller/RPi-Jukebox-RFID/blob/17b796294d503935b9c5c07b355074b37ca6fde0/src/jukebox/components/synchronisation/sync_shared/__init__.py)) but i am unsure now how to correctly integrate into the system.

1) I need a hook in the rfid reader and the play_card from playermpd.
To keep cross dependecies low i think a direct import is not wanted. So this left me with this options:
Make the calls via plugs rpc ([playermpd](https://github.com/AlvinSchiller/RPi-Jukebox-RFID/blob/17b796294d503935b9c5c07b355074b37ca6fde0/src/jukebox/components/playermpd/__init__.py#L452))
Or register as a callback ([rfidReader](https://github.com/AlvinSchiller/RPi-Jukebox-RFID/blob/17b796294d503935b9c5c07b355074b37ca6fde0/src/jukebox/components/rfid/reader/__init__.py#L184))
What would be the prefered solution on this?

2) How should the setup be integrated? A seperate script in the component like in v2? Or should this be part of the overall setup somehow? Or another option i am currently not aware of?

I would be thankful for any help and advice.

### CODE REVIEW

**Comment by pabera** [comment]
Hey Alvin. Sorry for not getting back to your earlier. Thanks for your contribution. Before we can go ahead, would you mind quickly summarizing what the feature does in V2? I am not aware of it and I can't really derive it from "synch shared". I will then be able to help with the integration into `future3`.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
+"""
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:1]: This is quite a long file. You could outsource helper functions into a separate file to optimize readability (see `components/rfid`)

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+Handles the synchronisation of shared audiofolder and the card database
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:2]: I am seeing a few redundant things, e.g. for me there shouldn't be a difference between the function `sync_card` and `sync_folder` 🤷‍♀️

I am curious.. is this a) a 2-way sync, b) Phoniebox -> Server or c) Server -> Phoniebox?

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+            if option == 'true':
+                _new_state = True
+            elif option == 'false':
+                _new_state = False
+            elif option == 'toggle':
+                _new_state = not self._sync_on_rfid_scan_enabled
+            else:
+                logger.error(f"Invalid value '{option}' for 'option' in 'sync_change_on_rfid_scan'")
+                _new_state = None
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:94]: A variable that can have a boolean state should only be boolean.. what kind of state is `True` compared to a state `toggle`? It's also not easy to read from the code what `True` or `False` or even `Toggle` mean

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        """
+        Sync full from the remote server
+        """
+        _files_synced = False
+
+        if self._sync_enabled:
+            logger.info("Syncing full")
+            _database_synced = self._sync_card_database()
+            _folder_synced = self._sync_folder('')
+            _files_synced = _database_synced or _folder_synced
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:116]: I'd expect this to be one single call, maybe with an array of paths (files are also just paths)

```
paths = [ path_to_db, path_to_audiofolder, path_to_something_else ]
for path in paths:
    self.sync(path)
```

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+        else:
+            logger.debug("Sync shared deactivated")
+
+        return _files_synced
+
+    @plugs.tag
+    def sync_card_database(self, card_id: str) -> bool:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:124]: This function almost looks like `sync_full` ;-)

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+            else:
+                logger.debug("Sync on RFID scan deactivated")
+
+        else:
+            logger.debug("Sync shared deactivated")
+
+        return _files_synced
+
+    @plugs.tag
+    def sync_folder(self, folder: str) -> bool:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:143]: same here ...

File: src/webapp/public/locales/de/translation.json [hunk 7]
```diff
@@ -58,6 +61,12 @@
         },
         "timers": {
           "description": "Wähle die Anzahl der Minuten nachdem die Aktion ausgeführt werden soll."
+        },
+        "synchronisation": {
+            "description": "Wähle den zu setzenden Status.",
+            "label-toggle": "Umschalten",
+            "label-true": "Aktivieren",
+            "label-false": "Deaktivieren"
```
Comment (pabera) [inline src/webapp/public/locales/de/translation.json:69]: Aah.. `True` means `activate` .. you should call it that way in your code as well

File: src/webapp/src/config.js [hunk 8]
```diff
@@ -30,41 +30,49 @@ const ROOT_DIRS = ['./', DEFAULT_AUDIO_DIR];
...
+        }
+    },
+
+    // Audio & Volume
+    audio: {
+        commands: {
+            change_volume: {},
+            toggle_output: {}
+        },
+    },
```
Comment (pabera) [inline src/webapp/src/config.js:49]: Let's try to avoid using 4 space indentation .. in JS we use 2 spaces.

**Review by pabera — CHANGES_REQUESTED** [review]

File: resources/default-settings/jukebox.default.yaml [hunk 9]
```diff
@@ -18,6 +18,7 @@ modules:
     host: hostif.linux
     bluetooth_audio_buttons: controls.bluetooth_audio_buttons
     gpio: gpio.gpioz.plugin
+    sync_shared: synchronisation.sync_shared
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply resources/default-settings/jukebox.default.yaml:21]: The idea of the feature was to sync elements in the "shared" folder.
I took the same name as in v2 (here "audiofolders" and "shortcuts" had been the only elements in shared), to be identified as the same feature.

I didn't choose "remote" or "from remote" in the first place, as it doesn't say what is synced and the direction may not always be only to the phoniebox (further additions to sync files back to the server are possible, yet not currently planned).

But we can try to find a name that fits it all.

File: resources/default-settings/sync_shared.default.yaml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+      server: ''
+      port:  # int
+      timeout: 1 # int
+      path: ''
+      username: ''
+  mount:
+      server: ''
+      port:  # int
+      timeout: 1 # int
+      path: ''
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply resources/default-settings/sync_shared.default.yaml:14]: True, i used it mainly to switch fast and test both cases , but this is probably no scenario in a real use.
I will change that.

File: src/jukebox/components/playermpd/__init__.py [hunk 11]
```diff
@@ -447,6 +447,16 @@ def play_card(self, folder: str, recursive: bool = False):
             self.second_swipe_action()
         else:
             logger.debug('Calling first swipe action')
+
+            # sync audiofolder
+            _files_synced = plugs.call_ignore_errors('sync_shared', 'ctrl', 'sync_folder', args=[folder])
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/playermpd/__init__.py:452]: I haven't found a better place for this, as the synchronisation is needed before the folder is actually played. 
If i would call this before playermpd is invoked (in e.g. the rfid_reader), i would have fiddle with the decoded [card_action](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/e4e718ec8c85d257397b5723042b6d31a3d25a98/src/jukebox/components/rfid/reader/__init__.py#L219) to see if play_card is invoked and which args is the folder name. And i personally see this as a nogo.

If the plugin isn't enabled (as in no 'modules' entry in jukebox.yaml) the error is ignored and playback is not harmed, but it would still be logged.
I think here comes my question 1 from the first post in place, on how to make the calls best.

File: src/webapp/src/config.js (hunk 8 above)
Comment (AlvinSchiller) [reply src/webapp/src/config.js:49]: That was enforced by the editorconfig settings. I will correct it.
Maybe this should be included in editorconfig to not be changed again on accident?

File: src/jukebox/components/playermpd/__init__.py [hunk 12]
```diff
@@ -585,6 +595,22 @@ def set_volume(self, volume):
...
         return self.get_volume()
 
+    def _wait_database_update_finished(self, update_id: int):
+        logger.debug("Waiting for update to finish")
+        with self.mpd_lock:
+            while self._is_database_updating(update_id):
+                # a little throttling
+                time.sleep(0.1)
+
+    def _is_database_updating(self, update_id: int):
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/playermpd/__init__.py:605]: done

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+            logger.error(f"Server not reachable: {_host}:{_port}. {e.__class__.__name__}: {e}")
+            return False
+
+        _server_reachable = result == 0
+        if not _server_reachable:
+            logger.error(f"Server not reachable: {_host}:{_port}. errorcode: {result}")
+
+        return _server_reachable
+
+    def _remote_is_file(self, path: str) -> bool:
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:282]: aggree. 
changed the name to _isfile_remote and _isdir_remote (naming like os.path functions)

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 2 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:2]: Currently its only c) Server -> Phoniebox.

They are not quite the same. The audiofolders are just placed in the file system and mpd can performe the update after.
The card database behaves a bit different as its a singel file for all cards, and therefore a special treatment is needed to update correctly (and only if changes are present). See also the comment in the header, where i tried to describe this specialities.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 3 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py]: changed the options to "enable" and "disable" intead "true" and "false" for better readability

File: src/webapp/public/locales/de/translation.json (hunk 7 above)
Comment (AlvinSchiller) [reply src/webapp/public/locales/de/translation.json]: changed the options to "enable" and "disable" intead "true" and "false" for better readability

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 5 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:121]: moved identical precheck to functions.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 6 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:133]: moved identical precheck to functions.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 4 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:116]: They are not quite the same. The audiofolders are just placed in the file system and mpd can performe the update after.
The card database behaves a bit different as its a singel file for all cards, and therefore a special treatment is needed to update correctly (and only if changes are present). See also the comment in the header, where i tried to describe this specialities.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+                cfg_sync_shared.setn('sync_shared', 'on_rfid_scan_enabled', value=_new_state)
+                self._sync_on_rfid_scan_enabled = _new_state
+
+                logger.info(f"Changed 'on_rfid_scan_enabled' to '{_new_state}'")
+
+        else:
+            logger.debug("Sync shared deactivated")
+
+    @plugs.tag
+    def sync_full(self) -> bool:
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:106]: sync_all sounds better. I will change that

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 1 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:1]: I will look into this

File: src/webapp/src/config.js (hunk 8 above)
Comment (pabera) [reply src/webapp/src/config.js]: It probably should be

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 4 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:117]: Just in case of "sync_all", where the card.yaml is just replaced.
But in case of "sync_on_scan" only the entry of this single CardId has to be updated. Otherwise it would break the offline functionality, as other cards could have been changed, but the audiofiles are not yet present on the phoniebox.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+        :param folder: Folder path relative to music library path
+        """
+        _files_synced = False
+
+        if self._precheck_enablement_sync_on_rfid_scan():
+            _files_synced = self._sync_folder(folder)
+
+        return _files_synced
+
+    def _precheck_enablement_sync(self) -> bool:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:155]: the wording here is a little clumsy. I'd call it it: `is_sync_enabled`

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+        if self._sync_enabled:
+            return True
+        else:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:158]: No need of else ;-)

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+            logger.debug("Sync shared deactivated")
+
+        return False
+
+    def _precheck_enablement_sync_on_rfid_scan(self) -> bool:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:163]: Same here, update naming: `is_sync_enabled_on_rfid_scan`

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 18]
```diff
@@ -0,0 +1,368 @@
...
+        if self._precheck_enablement_sync():
+            if self._sync_on_rfid_scan_enabled:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:165]: You could write this more concise

```python
if self.is_sync_enabled and self.sync_on_rfid_scan_enabled:
  return True

logger.debug("Sync on RFID scan deactivated")
return False
```

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+                logger.debug("Sync on RFID scan deactivated")
+
+        return False
+
+    def _sync_card_database(self, card_id: str = None) -> bool:
+        _card_database_path = cfg_cards.loaded_from
+        logger.info(f"Syncing card database: {_card_database_path}")
+        _files_synced = False
+
+        if self._is_server_reachable():
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:177]: I wouldn't wrap this all in one block, you could decouple it like this

```python
if not self._is_server_reachable():
  ...
  return _files_synced # or immediately `False`

_sync_remote_path_settings = ....
...
_dst_path = os.path.join(_card_database_dir, "sync_temp_" + _card_database_file)

if self._isfile_remote(_src_path):
  _files_synced = self._sync_paths(_src_path, _dst_path)
  ...
```

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 20]
```diff
@@ -0,0 +1,368 @@
...
+            else:
+                logger.warn(f"Card database does not exist remote: {_src_path}")
+
+        return _files_synced
+
+    def _sync_folder(self, folder: str) -> bool:
+        logger.info(f"Syncing Folder '{folder}'")
+        _files_synced = False
+
+        if self._is_server_reachable():
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:221]: same here..

```python
if not self._is_server_reachable():
  return False

...
```

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 21]
```diff
@@ -0,0 +1,368 @@
...
+            logger.error(f"Server not reachable: {_host}:{_port}. {e.__class__.__name__}: {e}")
+            return False
+
+        _server_reachable = result == 0
+        if not _server_reachable:
+            logger.error(f"Server not reachable: {_host}:{_port}. errorcode: {result}")
+
+        return _server_reachable
+
+    def _isfile_remote(self, path: str) -> bool:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:295]: `_is_file_remote` .. I would separate `is` and `file`

File: src/jukebox/components/synchronisation/sync_shared/__init__.py [hunk 22]
```diff
@@ -0,0 +1,368 @@
...
+                                shell=False, check=False, capture_output=True, text=True)
+
+            _result = _runresult.returncode == 0
+
+        else:
+            _result = os.path.isfile(path)
+
+        return _result
+
+    def _isdir_remote(self, path: str) -> bool:
```
Comment (pabera) [inline src/jukebox/components/synchronisation/sync_shared/__init__.py:313]: `_is_dir_remote`.. same here

File: resources/default-settings/jukebox.default.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (pabera) [reply resources/default-settings/jukebox.default.yaml:21]: I thought about it.. why don't you just call it `filesync` .. and the file structure could be like this

`filesync/remote` instead of `synchronisation/sync_shared`

in your `jukebox.yaml` you could register it this way

```yaml
  filesync: filesync.remote
```

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 15 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:155]: changed name to "_is_sync_enabled"

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 17 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:163]: changed name to "_is_sync_enabled"

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 16 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:158]: removed unneccessary "else"

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 18 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:165]: For a more precise logging i would keep the seperation.
Otherwise if "is_sync_enabled" is false there would still be the log entry for "rfid scan" even if this setting might be activated, which could case confusion.

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 21 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py:295]: done

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 19 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py]: done

File: src/jukebox/components/synchronisation/sync_shared/__init__.py (hunk 20 above)
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/sync_shared/__init__.py]: done

File: src/jukebox/components/rfid/reader/__init__.py [hunk 23]
```diff
@@ -38,7 +39,10 @@ def register(self, func: Callable[[str, int], None]):
             :noindex:
 
             :param card_id: Card ID
-            :param state: 0 if card id is registered, 1 if card id is unknown
+            :param state:
+                -1  on receiving card id (before processing card entries),
+                0   if card id is registered,
+                1   if card id is unknown
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/rfid/reader/__init__.py]: Refactored this to have the state as enum to have defined entries (instead of magic string, int, ...).

Used this also in playermpd with the addition that the Callback class has now a generic for state which is given on construction. This way only a new state enum must be added to add a new callback for different play-functions (but maybe just to much foreseeing?)

What do you think about this solution?

File: resources/default-settings/jukebox.default.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply resources/default-settings/jukebox.default.yaml]: changed name to "synchronisation/rfidcards" ("sync_rfidcards" in jukebox.yaml)

"synchronisation" is the generic component, "rfidcards" the specific implementation. 
For now RFID card data (audiofolders and card database entries) are the only synced items.

File: docs/sphinx/userguide/sync_rfidcards.rst [hunk 24]
```diff
@@ -0,0 +1,71 @@
+Syncronisation RFID Cards
```
Comment (pabera) [inline docs/sphinx/userguide/sync_rfidcards.rst:1]: I made a few English tweaks here and there and directly committed the changes.

File: docs/sphinx/userguide/sync_rfidcards.rst [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+For the "sync-all" bind a cardId to the command.
+For the "RFID scan sync" feature, activate the option in the configuration
+or bind a cardId to the command for dynamic de-/activation.
+
+Synchronisation
+---------------
+
+The synchronisation will be FROM a server TO the Phoniebox, overriding existing files.
+So configuration made locally will be lost after sync.
+If you want to make the initial setup e.g. via WebUi copy the files and use it as a base for the server.
```
Comment (pabera) [inline docs/sphinx/userguide/sync_rfidcards.rst:20]: I didn't quite get what you are trying to say here :-)
Maybe you can break it into 2 sentences.

**Review by pabera — APPROVED** [review]

File: src/jukebox/components/synchronisation/rfidcards/__init__.py [hunk 26]
```diff
@@ -0,0 +1,358 @@
+"""
+Handles the synchronisation of RFID cards (audiofolder and card database entries).
+
+sync-all -> all card entries and audiofolders are synced from remote including deletions
+sync-on-scan -> only the entry and audiofolder for the cardId will be synced from remote.
+                Only deletions of subelements will be performed
```
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/rfidcards/__init__.py:6]: A generalisation for files or subfolder that exists inside the linked folder, which will be affected by deletion.
But a deletion of the linked folder itself on remote side will not be propagated.

File: docs/sphinx/userguide/sync_rfidcards.rst (hunk 25 above)
Comment (AlvinSchiller) [reply docs/sphinx/userguide/sync_rfidcards.rst:20]: It was just a hint if someone wants to make the initial configuration of rfidcards or audiofolder via webui, there is no way to sync them TO the remote server.
These files need to be copied manually.

Maybe rephrase to:
The synchronisation will be only TO the Phoniebox, overriding existing files and configuration if present.

File: src/jukebox/components/synchronisation/rfidcards/__init__.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (pabera) [reply src/jukebox/components/synchronisation/rfidcards/__init__.py:6]: Maybe we should write that way in the docs ;-)

[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply src/jukebox/components/synchronisation/rfidcards/__init__.py]: Updated the docs

**Comment by pabera** [comment]
Well done. Thanks a lot for the effort. Merging now.
